gapir/cc/resource_cache.cpp

```diff
@@ -23,51 +23,116 @@
 #include <vector>
 
 namespace gapir {
-size_t ResourceCache::prefetch(const Resource* res, size_t count,
-                               ResourceLoader* fetcher) {
-  size_t res_sum = 0;
-  std::vector<Resource> uncached;
-  uncached.reserve(count);
-  size_t alreadyCached = 0;
+
+size_t ResourceCache::setPrefetch(const Resource* resources, size_t count,
+                                  std::unique_ptr<ResourceLoader> fetcher) {
+  mResources.clear();
+  mResourceIterators.clear();
+
+  mResources.reserve(count);
+
+  for (unsigned int i = 0; i < count; ++i) {
+    mResources.push_back(resources[i]);
+    mResourceIterators[resources[i].getID()] = mResources.end() - 1;
+  }
+
+  mFetcher = std::move(fetcher);
+
+  return prefetchImpl(resources, count);
+}
+
+std::vector<Resource> ResourceCache::anticipateNextResources(
+    const Resource& resource, size_t bytesToFetch) {
+  std::vector<Resource> expectedResources;
+
+  size_t bytesSoFar = 0;
+
+  auto resMapIter = mResourceIterators.find(resource.getID());
+  if (resMapIter == mResourceIterators.end()) {
+    return expectedResources;  // We don't know about this resource so we're
+                               // blind. Return the empty vector.
+  }
+
+  std::vector<Resource>::iterator resIter = resMapIter->second;
+
+  if (resIter != mResources.end()) {
+    ++resIter;
+  }
+
+  for (unsigned int i = 0;
+       resIter != mResources.end() && bytesSoFar < bytesToFetch; ++i) {
+    expectedResources.push_back(*resIter);
+    bytesSoFar += resIter->getSize();
+    resIter++;
+  }
+
+  return expectedResources;
+}
+
+size_t ResourceCache::prefetchImpl(const Resource* resources, size_t count,
+                                   bool allowEviction) {
+  size_t bytesToFetch = 0;
+
+  std::vector<Resource> uncachedResources;
+  uncachedResources.reserve(count);
+
+  size_t numResourcesAlreadyCached = 0;
 
   for (size_t i = 0; i < count; i++) {
-    const auto& r = res[i];
-    if (hasCache(r)) {
-      alreadyCached++;
+    const auto& resource = resources[i];
+
+    if (hasCache(resource)) {
+      numResourcesAlreadyCached++;
       continue;
     }
-    if (res_sum + r.size > totalCacheSize()) {
+
+    if (bytesToFetch + resource.getSize() >
+        (allowEviction ? totalCacheSize() : unusedSize())) {
       break;
     }
-    uncached.push_back(r);
-    res_sum += r.size;
+
+    uncachedResources.push_back(resource);
+    bytesToFetch += resource.getSize();
   }
+
+  ResourceLoadingBatch bat;
```

Comment:
Can you rename this variable to `batch` , and move its declaration under the `GAPID_INFO()` call which does not refer to it?